### ISSUE TITLE
Premultiplied alpha sprites do not blend correctly

### ISSUE THREAD

**Issue by jabuwu** [bevyengine/bevy#6315]
## Bevy version

0.8.1

## What you did

I've been working on bringing the [Spine](http://esotericsoftware.com/) runtime to Bevy and one of the export options is "Premultiplied Alpha." This is a common rendering technique to solve sampling issues, especially when resizing sprites at runtime. The only requirement to work with premultiplied sprites is to change the blend mode. There's a blend mode [exported by Bevy](https://docs.rs/bevy/0.8.1/bevy/render/render_resource/struct.BlendState.html#associatedconstant.PREMULTIPLIED_ALPHA_BLENDING) for this exact thing.

## What went wrong

The blending is not applied correctly. It is particularly noticeable when overlapping the same color, as a black outline appears. In the screenshot below, the eye is a separate sprite rendered on top of the head.

<img width="328" alt="Black Outline" src="https://user-images.githubusercontent.com/46233424/196840586-c6837465-7b3e-41a0-a53f-b9079ebaa349.png">

## Additional information

I have created an [example repo](https://github.com/jabuwu/bevy_pma) which not only demonstrates the problem, but also shows a workaround, which might help narrow down the cause.

The workaround is [a correction shader](https://github.com/jabuwu/bevy_pma/blob/main/assets/corrected.wgsl) which converts the texture into nonlinear space, undoes the premultiplied alpha, converts back to linear space, and multiplies the alpha again.

I've also created [an image generation function](https://github.com/jabuwu/bevy_pma/blob/main/src/image_gen.rs#L14) which can apply premultiplication in either SRGB space or Linear space. Bevy correctly renders a sprite that is premultiplied in linear space, but not in SRGB space.

I've tested these sprites in Unity and a simple webgl example ([which I've included in the repo](https://github.com/jabuwu/bevy_pma/tree/main/html)), and Bevy behaves differently than both of them. Note that the html example does set a "premultipliedAlpha" option to true, but this is only so that it can blend with the background, and setting this to false does not affect how the spineboy sprites blend with each other.

All of this seems to suggest to me that the alpha channel of the images are being changed when imported by Bevy. I tried looking into this myself but haven't been able to track down the problem yet.

**Comment by sbaudlr** [comment]
I've been taking a look into this. Just to note: PNGs should not have premultiplied alpha ([source](https://www.w3.org/TR/png/#6AlphaRepresentation)) so the outline you're seeing is behaving "as expected" in that regard.

However, I've done some testing using TARGA files with premultiplied alpha (sorry I can't share them - I'll try to make some that I can share) and there does seem to be an issue with applying the alpha correctly, so there is _a_ bug here I believe.

I suspect that when I track down that issue I might find some related code that could allow your images to work - question is whether PNGs with premultiplied alpha should be supported as this is off-spec, but to my understanding this is something other game engines do allow.

**Comment by jabuwu** [comment]
Thanks for looking into it @sbaudlr! Yes, sorry, I should have mentioned I tried other formats than PNG. However, I don't think it's "as expected," at least with my understanding of the problem. My simple [WebGL example](https://github.com/jabuwu/bevy_pma/blob/main/html/index.html) with these same images blends correctly. I think whatever is causing the problem generally is the same problem with PNG, and there's nothing special that needs to be supported with off-spec PNGs.

If I end up being wrong here, maybe it can be properly addressed with asset preprocessing as mentioned in the Bevy 0.9 release post.

**Comment by sbaudlr** [comment]
Yeah poor choice of words on my part, it's unsurprising to me that a PNG with premultiplied alpha is not showing correctly.

Your WebGL example works because you're telling WebGL to treat it like premultiplied alpha which makes sense. So far I've convinced myself that webgpu is not at fault, neither is anything after the material is created (I could be wrong though!). Which points in the direction of the asset loading being at fault. Could be something to do with converting the image into the format the `bevy_render` wants material textures to be in that's specific to loading from images rather than generating color procedurally like in your example.

Speaking of which, the reason the `LINEAR` option in your example works whereas `SRGB` doesn't is because you must apply alpha in linear color space - before gamma correction.

**Comment by jabuwu** [comment]
> Your WebGL example works because you're telling WebGL to treat it like premultiplied alpha which makes sense.

I'm telling Bevy to treat it as premultiplied alpha too. I'm only mentioning it because in all my reading of the topic, the only requirement for PMA is to set the proper blend mode. I understand that the conversion between SRGB and Linear which happens throughout the rendering process complicates things. To be honest, that's the most confusing part for me when trying to hunt this down myself.

If you are referring to the `premultipliedAlpha` setting in WebGL, setting it to true allows proper blending with the background, and has nothing to do with how the sprites blend with each other, nor how the images are imported. The images still blend correctly (with each other) if the value is false.

> Could be something to do with converting the image into the format the `bevy_render` wants material textures to be in

Yes, I think this as well!

**Comment by sbaudlr** [comment]
Aye, your WebGL example is what originally convinced me that it must be during loading, if a premultiplied image makes it to render-time unscathed then it works 😄

**Comment by sbaudlr** [comment]
I think I know part of what's going on.

So, images in bevy are converted to linear color before being loaded onto the GPU, which is correct as far as color manipulation is concerned. If I change this hard-coded (boo) value from `true` to `false` in [image_texture_loader](https://github.com/bevyengine/bevy/blob/v0.9.0/crates/bevy_render/src/texture/image_texture_loader.rs#L53) then I end up with this result:

![bevy_pma_lie_to_bevy](https://user-images.githubusercontent.com/107637688/204904643-aefe2b7d-e1c2-4b14-a3d1-eff3e5a1f0a6.png)

The alpha blending has produced the right result (no more border around the eye texture). However, because we lied to bevy and told it that a gamma-corrected image should be treated as linear color, bevy has then applied gamma correction to the result, resulting in everything looking washed out - most notably the hair.

This works for the same reason that the WebGL example works: no gamma adjustment is being made before compositing the two images - which brings me to the issue.

As your corrected shader demonstrates, the alpha has been pre-multiplied in the source image _after_ gamma correcton has been applied. This is not correct, it _must_ be done before gamma correction for exactly the reason that is demonstrated here.

Bevy converts your image to linear space for compositing, which also manipulates the alpha-multiplied pixels. This means that when the linear-space images are composited together, the compositing is performed on "darkened" pixels. If the alpha is applied before gamma correction, then the right alpha value is maintained when the image is brought back to linear space.

To prove this to myself, I [wrote some code](https://github.com/sbaudlr/bevy_6315_correct_premultiplied_alpha) that:
1. Loads your image.
2. Divides by alpha in gamma-corrected space.
3. Converts your image to linear space.
4. Applies the alpha.
5. Applies the gamma curve.
6. Saves the corrected image as a new file.

If I load this new version of the asset into your example (with no modifications to bevy) then the "vanilla" version looks correct! And you can see the "corrected" version now looks overblown:

![bevy_pma_corrected_alpha](https://user-images.githubusercontent.com/107637688/204904695-b666bc98-a990-4f4c-a881-32581c08fe41.png)

I don't know why this works in Unity / Unreal, perhaps they are compositing the images in gamma-corrected space? I would hope not...

Likewise, WebGL is perfectly happy to just do the maths you've told it to because it trusts that you're feeding it the right kind of data.

As for the issues I have been seeing, they look to me like scaling issues - scaling images with pre-multiplied alpha is tricky. But notably the artifacts caused are small white lines, rather than dark lines, which is a pretty good indicator for them being caused by a different category of premultiplied alpha bug - so I'll look into that and open a separate issue if I find anything of interest there.

I'd be interested to know what process you're using to generate the premultiplied image? It could be that switching the order of two steps is all that's needed. But I don't think this is an issue with bevy, other than _maybe_ there not being a way to perform compositing in gamma-corrected space (again, compositing should always be done in linear space!)

Apologies for the long-winded write-up, I wanted to show my entire thought process because I could be wrong! Maybe you can spot a flaw in my process?

**Comment by jabuwu** [comment]
Thank you for the writeup. I mostly agree with you and I've done a lot of these same steps in trying to hunt down the problem, but ran into the same confusion as you: why does this work in other engines and WebGL?

If the answer really is: everyone is doing it wrong and Bevy is doing it right, then I'm happy to accept that. For my purposes, I want [bevy_spine](https://github.com/jabuwu/bevy_spine) to work without any hassle. Certainly not introducing an extra step to fix the images like you had to. Currently my fix is the shader hack but I can probably find a better solution (like automating those fixes you did when importing images for spine, rather than at runtime in the shader).

> I'd be interested to know what process you're using to generate the premultiplied image?

I'm using Spine which creates a sprite atlas and has an option for premultiplied alpha. Maybe there's another setting in here that I should be using? But, this is the default, and given the popularity of Spine I assume they're doing it correctly. I acknowledge that might be a bad assumption.

<img width="659" alt="Screen Shot 2022-11-30 at 2 14 09 PM" src="https://user-images.githubusercontent.com/46233424/204919549-1881d3b7-b8d0-4ae0-8602-7d2402aa7e40.png">

> As for the issues I have been seeing, they look to me like scaling issues - scaling images with pre-multiplied alpha is tricky.

I'm not sure what you're referring to here, but it was my understanding that premultiplied alpha makes scaling images more accurate. It's the problem that PMA fixes since [scaling non-PMA images is problematic](https://developer.nvidia.com/content/alpha-blending-pre-or-not-pre).

**Comment by viridia** [comment]
Suggestion: Now that Bevy asset V2 has landed, it should be possible to make "premultiplied alpha" a configuration parameter for the image loader, which could either be specified directly in code for a given asset handle, or in an associated .meta file for that asset file.

This doesn't address the issue of whether premultiplied alpha is valid for PNG, but it would let you punt on the issue by overriding the default behavior.
